### PR TITLE
Dynamic role update

### PR DESCRIPTION
**Pull request by emrahkocaman**
API changes for dynamically updating client permission configs.

### CODE REVIEW

File: hazelcast/src/main/java/com/hazelcast/osgi/impl/HazelcastOSGiInstanceImpl.java [hunk 1]
```diff
@@ -50,6 +50,7 @@
 import com.hazelcast.quorum.QuorumService;
 import com.hazelcast.ringbuffer.Ringbuffer;
 import com.hazelcast.scheduledexecutor.IScheduledExecutorService;
+import com.hazelcast.security.SecurityService;
```
Comment (jerrinot) [inline hazelcast/src/main/java/com/hazelcast/osgi/impl/HazelcastOSGiInstanceImpl.java]: is this passing checkstyle? it looks unsed?

Comment (emrahkocaman) [inline hazelcast/src/main/java/com/hazelcast/osgi/impl/HazelcastOSGiInstanceImpl.java]: It's not the only one, cleanup is on the way

File: hazelcast/src/main/java/com/hazelcast/internal/dynamicconfig/DynamicConfigurationAwareConfig.java [hunk 2]
```diff
@@ -1070,7 +1074,9 @@ public Config setServicesConfig(ServicesConfig servicesConfig) {
 
     @Override
     public SecurityConfig getSecurityConfig() {
-        return staticConfig.getSecurityConfig();
+        DynamicSecurityConfig securityConfig = new DynamicSecurityConfig(staticConfig.getSecurityConfig());
```
Comment (mdogan) [inline hazelcast/src/main/java/com/hazelcast/internal/dynamicconfig/DynamicConfigurationAwareConfig.java]: `securityService` can be injected into constructor directly.

File: hazelcast/src/main/java/com/hazelcast/internal/dynamicconfig/DynamicSecurityConfig.java [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+ * {@link SecurityConfig} wrapper that supports dynamically updating
+ * client permissions.
+ *
+ * @see #setClientPermissionConfigs(Set)
+ */
+public class DynamicSecurityConfig extends SecurityConfig {
+
+    private final  SecurityConfig securityConfig;
+
+    private SecurityService securityService = new NoOpSecurityService();
```
Comment (mdogan) [inline hazelcast/src/main/java/com/hazelcast/internal/dynamicconfig/DynamicSecurityConfig.java]: `securityService` can be injected via constructor directly and made `final`.

File: hazelcast/src/main/java/com/hazelcast/security/IPermissionPolicy.java [hunk 4]
```diff
@@ -46,6 +48,8 @@
      */
     PermissionCollection getPermissions(Subject subject, Class<? extends Permission> type);
 
+    boolean refreshClientPermissions(Set<PermissionConfig> permissionConfigs);
```
Comment (mdogan) [inline hazelcast/src/main/java/com/hazelcast/security/IPermissionPolicy.java]: Why not `refreshPermissions()`? `getPermissions()` method doesn't contain `client`.

File: hazelcast/src/main/java/com/hazelcast/security/SecurityContext.java [hunk 5]
```diff
@@ -104,4 +107,6 @@ void interceptBefore(Credentials credentials, String serviceName, String objectN
      */
     void destroy();
 
+    void refreshClientPermissions(Set<PermissionConfig> permissionConfigs);
```
Comment (mdogan) [inline hazelcast/src/main/java/com/hazelcast/security/SecurityContext.java]: Similar to `IPermissionPolicy`, why not just `refreshPermissions()`?

Comment (emrahkocaman) [inline hazelcast/src/main/java/com/hazelcast/security/SecurityContext.java]: `SecurityConfig` sets these permissions via `setClientPermissionConfigs` method so I've also used a similar term.

File: hazelcast/src/main/java/com/hazelcast/security/IPermissionPolicy.java (hunk 4 above)
Comment (emrahkocaman) [inline hazelcast/src/main/java/com/hazelcast/security/IPermissionPolicy.java]: Same reason as  `SecurityContext` but `IPermissionPolicy` doesn't refer `client` anywhere else so I'll change this. (But not sure about changing `SecurityContext`)

Comment (emrahkocaman) [inline hazelcast/src/main/java/com/hazelcast/security/IPermissionPolicy.java]: Uh ok, it also doesn't use `client` prefix. I'll change that too.

**Comment by emrahkocaman** [comment]
Seems like an unrelated failure

**Comment by emrahkocaman** [comment]
run-lab-run

File: hazelcast/src/main/java/com/hazelcast/internal/management/dto/PermissionConfigDTO.java [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+        out.writeUTF(permissionConfig.getType().getNodeName());
+        out.writeUTF(permissionConfig.getName());
+        if (StringUtil.isNullOrEmptyAfterTrim(permissionConfig.getPrincipal())) {
+            out.writeUTF("*");
+        } else {
+            out.writeUTF(permissionConfig.getPrincipal());
+        }
+
+        Set<String> endpoints = permissionConfig.getEndpoints();
+        if (endpoints != null) {
```
Comment (vbekiaris) [inline hazelcast/src/main/java/com/hazelcast/internal/management/dto/PermissionConfigDTO.java]: `permissionConfig.getEndpoints` works with get-or-create semantics so the `null` check here is redundant. It's the same for `permissionConfig.getActions` below also.

**Comment by vbekiaris** [comment]
Continued & merged in #10832

**Pull request closed without merging**